File: MatlabCode/examples.m
%% Set up
clear; close all
q = shiftOperator('q');
one = shiftOperator(1);
z = shiftOperator(0);
r = 0.5;                % Discount factor

%% Introduction example
% Matrix representing the transportation routs and delays
m = 4;
p = m+1;
M = shiftOperator.zeros(p,m);
for i = 1:m
    M(i,i) = q'*r;
    M(i+1,i) = -one;
end

% plotGraph(M);     % Uncomment if you want to plot the graph


% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr_formula(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

%% Simple example
% Matrix representing the transportation routs and delays
M = [q'*r z z;
    -one q'*r z;
    z -one q'*r;
    z z -one];

% plotGraph(M);     % Uncomment if you want to plot the graph

[p,m] = size(M);

% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr_formula(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

%% Tree 1
% Matrix representing the transportation routs and delays
M = [q'*r z z z;
    z q'*r z z;
    -one z q'*r z;
    z -one -one q'*r;
    z z z -one];

% plotGraph(M);     % Uncomment if you want to plot the graph

[p,m] = size(M);

% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr_formula(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

%% Tree 2
% Matrix representing the transportation routs and delays
M = [q'*r z z z;
    z -one z z;
    -one z -one z;
    z q'*r q'*r q'*r;
    z z z -one];

% plotGraph(M);     % Uncomment if you want to plot the graph

[p,m] = size(M);

% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr_formula(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

%% String
% Matrix representing the transportation routs and delays
M = [-one z z z;
    q'*r -one z z;
    z q'*r q'*r z;
    z z -one q'*r;
    z z z -one];

% plotGraph(M);     % Uncomment if you want to plot the graph

[p,m] = size(M);

% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

%% Tree 3
rng(2)

% Creat random matrix representing the transportation routs and delays
m = 20;
mixedDirection = false;
outflow = -one;
inflow = q'*r;
M = randomTree(m, mixedDirection, outflow, inflow);

% plotGraph(M);     % Uncomment if you want to plot the graph

[p,m] = size(M);


% Get the state space representation
sys = M2ss(M);
A = sys.A;
C = sys.C;
B = sys.B;

% Find control law
[K1, K2] = factor_lqr(A, C, M, r)

% Check validity
K = K1\K2;
Q = C'*C;
R = zeros(m);
[~, K_idare, ~] = idare(A*r,B,Q,R);
Kerror = K - K_idare

File: JuliaCode/examples.jl
include("ShiftOperator.jl")
include("factor_lqr_formula.jl")
include("factor_lqr.jl")
include("M2ss.jl")
include("randomTree.jl")
using DelimitedFiles                # Required to load the large tree example
using Graphs, Plots, GraphRecipes   # Required if you wan to plot the graphs
include("plotGraph.jl")             # Required if you wan to plot the graphs

q = ShiftOperator("q")
id = ShiftOperator(1)
z = ShiftOperator(0)
r = 0.5
outflow_pattern = -id
inflow_pattern = q'*r

#------------------- Simple example -------------------------
# Matrix representing the transportation routs and delays
M = [q'*r z z;
     -id q'*r z;
     z -id q'*r;
     z z -id]
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr_formula(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")

#------------------- Tree 1 -------------------------
# Matrix representing the transportation routs and delays
M = [q'*r z z z;
    z q'*r z z;
    -id z q'*r z;
    z -id -id q'*r;
    z z z -id]
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr_formula(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")

#------------------- Tree 2 -------------------------
# Matrix representing the transportation routs and delays
M = [q'*r z z z;
    z -id z z;
    -id z -id z;
    z q'*r q'*r q'*r;
    z z z -id]
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr_formula(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")

#------------------- String -------------------------
# Matrix representing the transportation routs and delays
M = [-id z z z;
    q'*r -id z z;
    z q'*r q'*r z;
    z z -id q'*r;
    z z z -id]
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")

#------------------- Big  tree -------------------------
# The big tree in the artikel was generated with the randomTree function
# in Matlab. We provide an extra example bellow that uses Julias randomTree
# function.
M0 = readdlm("M0-big-tree.txt",',','\n')
M1 = readdlm("M1-big-tree.txt",',','\n')
M = M0*id + M1*q'
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")

#------------------- Random  tree -------------------------
m = 10
mixedDirections = false

M = randomTree(m, mixedDirections, outflow_pattern, inflow_pattern)
#plotGraph(M, outflow_pattern, inflow_pattern)   # Uncomment to plot the graph

p,m = size(M)
A,B,C,D = M2ss(M)

K1, K2 = factor_lqr(A,C,M,r)
display(K1)
print("\n")
display(K2)
print("\n")
print("\n")
